### PR TITLE
Suggest non-gki kernel users to backport path_umount

### PR DESCRIPTION
**Pull request by backslashxx**
Most kernel builders have to touch their kernel source code anyway, why not also tell them to backport path_umount so even non-gki users can benefit from this?

I know this might be a bit controversial as this will raise the barrier of entry, but the benefits are just so high.

Idea was from OnlyTomInSecond on KernelSU group chat way back, and it has been on the discussions for some time

references:
https://t.me/KernelSU_group/27237/176515
https://t.me/KernelSU_group/3249/184908
https://github.com/tiann/KernelSU/discussions/955#discussioncomment-7617166
https://github.com/OnlyTomInSecond/android_kernel_xiaomi_sdm845/commit/03d233db8bbe9fe101509430bfa09d1899168aa7
https://github.com/tiann/KernelSU/pull/1060 

https://elixir.bootlin.com/linux/v5.9.1/source/fs/namespace.c#L1728
https://elixir.bootlin.com/linux/v5.10.9/source/fs/namespace.c#L1730
https://elixir.bootlin.com/linux/v6.5/source/fs/namespace.c#L1887 https://github.com/tiann/KernelSU/pull/1464#issuecomment-2002492107

Kernel side change examples
5.4 https://github.com/natsumerinchan/kernel_oneplus_sm8350/commit/961d9788624e88c3c58918b3781b9fbdd19ecbeb 
4.19 https://github.com/backslashxx/android_karnol_ximi_fog/commit/164917f56d0e75ab51bb9f1bdf489acac7a6d3db
4.14 https://github.com/crdroidandroid/android_kernel_xiaomi_sm6150/commit/c07c70a0c59fb46ac0c1c812c3e993022e375cd5
4.9 https://github.com/backslashxx/msm8953-los21/commit/195f07593ae9769a992f3945bcdc43ff7fc99827
4.4 https://github.com/riarumoda/android_kernel_samsung_a9y18qlte/commit/21ea33fe41ce079ec1d663c0bd2201bc00a8084a https://github.com/tiann/KernelSU/pull/1464#issuecomment-2002424069
ofcourse having someone on 3.18 confirm this will be nice.

**PROS**: umount modules for everyone
**CONS**: barrier of entry +1

### CODE REVIEW

**Comment by Asyanx** [comment]
+1 work

**Comment by xennin** [comment]
pro

**Comment by Fede2782** [comment]
+1. It could really  uniform things up. I can confirm works on my 4.9 and 4.14

**Comment by alternoegraha** [comment]
+1

**Comment by Dwyor-tmx** [comment]
+1 work on my kramel 😋

**Comment by Dwyor-tmx** [comment]
+1 work on my kramel 😋

**Comment by Dwyor-tmx** [comment]
+1

**Comment by Dwyor-tmx** [comment]
> Most kernel builders have to touch their kernel source code anyway, why not also tell them to backport path_umount so even non-gki users can benefit from this?
> 
> I know this might be a bit controversial as this will raise the barrier of entry, but the benefits are just so high.
> 
> Idea was from OnlyTomInSecond on KernelSU group chat way back, and it has been on the discussions for some time
> 
> references: https://t.me/KernelSU_group/27237/176515 https://t.me/KernelSU_group/3249/184908 [#955 (comment)](https://github.com/tiann/KernelSU/discussions/955#discussioncomment-7617166) [OnlyTomInSecond/android_kernel_xiaomi_sdm845@03d233d](https://github.com/OnlyTomInSecond/android_kernel_xiaomi_sdm845/commit/03d233db8bbe9fe101509430bfa09d1899168aa7)
> 
> Kernel side change examples 5.4 [dev-sm8350/kernel_oneplus_sm8350@40fe07b](https://github.com/dev-sm8350/kernel_oneplus_sm8350/commit/40fe07bbc851b50c204c6a91128b5cbefb7a6296) [natsumerinchan/kernel_oneplus_sm8350@961d978](https://github.com/natsumerinchan/kernel_oneplus_sm8350/commit/961d9788624e88c3c58918b3781b9fbdd19ecbeb) 4.19 [backslashxx/android_karnol_ximi_fog@164917f](https://github.com/backslashxx/android_karnol_ximi_fog/commit/164917f56d0e75ab51bb9f1bdf489acac7a6d3db) 4.14 [backslashxx/mojito_krenol@686e90b](https://github.com/backslashxx/mojito_krenol/commit/686e90b6a63ad0851e964bddab7add54ec52b973) 4.9 [backslashxx/msm8953-los21@195f075](https://github.com/backslashxx/msm8953-los21/commit/195f07593ae9769a992f3945bcdc43ff7fc99827) ofcourse having someone on a 4.4 kernel confirm this will be nice.
> 
> TL;DR PROS: umount modules for everyone CONS: barrier of entry +1

+1 Work on my kramel 😋

**Comment by Dwyor-tmx** [comment]
+1

**Comment by privacyguy123** [comment]
+1

**Comment by Fede2782** [comment]
@backslashxx I don't know if there is a way to do this but the best way would be to dynamically enable/disable modules umount feature if there is the path_umount code in fs/namespace.c, because very likely @tiann will close this issue as he did with the other one

**Comment by FerryAr** [comment]
+1

**Comment by riarumoda** [comment]
Hi, 4.4 kernel user in here, i used the 4.9 patch and it works. my messy commit is in here:
https://github.com/riarumoda/android_kernel_samsung_a9y18qlte/commit/046ec1c9482a4b6828e112dcfd2966d1cd765129
https://github.com/riarumoda/android_kernel_samsung_a9y18qlte/commit/d370283ea8744cdb989d786404cd3acc6237bb44
https://github.com/riarumoda/android_kernel_samsung_a9y18qlte/commit/21ea33fe41ce079ec1d663c0bd2201bc00a8084a

**Comment by Fede2782** [comment]
> > @backslashxx I don't know if there is a way to do this but the best way would be to dynamically enable/disable modules umount feature if there is the path_umount code in fs/namespace.c, because very likely @tiann will close this issue as he did with the other one
> 
> is there a way to test a function if it exists before linking? if so then thats a plausible way to do it.

I would say to make that little part of code separated and it could be built or not by a Makefile statement if it locates the code in fs/namespace.c. I will try to implement this if i have time in these days

**Comment by Fede2782** [comment]
> > I would say to make that little part of code separated and it could be built or not by a Makefile statement if it locates the code in fs/namespace.c. I will try to implement this if i have time in these days if it is possible
> 
> Good idea. --check latest change I continue with #error warning and still forcing builder to do it, or of course, do nothing?? b-b-b-but I want everyone to do it bruv

Yes i think it could work, good job. I'll try to build later and see if works. In any case you should not force this change i think. Let the devs choose

**Comment by Fede2782** [comment]
I would prefer to use something like this but i am not sure. 
```c
static void ksu_umount_mnt(struct path *path, int flags)
{
#ifdef KSU_UMOUNT
	int err = path_umount(path, flags);
	if (err) {
		pr_info("umount %s failed: %d\n", path->dentry->d_iname, err);
	}
#else
	// TODO: umount for non GKI kernel
        #warning Did you know you can backport path_umount to fs/namespace.c from 5.9?
        #warning https://kernelsu.org/guide/how-to-integrate-for-non-gki.html#backport-path_umount-from-linux-59

#endif
}
```

File: website/docs/guide/how-to-integrate-for-non-gki.md [hunk 1]
```diff
@@ -292,6 +293,53 @@ index 45306f9ef247..815091ebfca4 100755
  		add_input_randomness(type, code, value);
 ```
 
+## Backport path_umount from Linux 5.9
+
+```
```
Comment (Fede2782) [inline website/docs/guide/how-to-integrate-for-non-gki.md:298]: You should add "diff" in the start of the code block and explain why is required. Maybe also write in the KPROBES section and manual edit section to read the backport path umount part

```
In order to make a more standard KernelSU implementation, you should backport the ```code_umount``` feature in ```fs/namespace.c``` if your kernel version is 5.9 or lower. Here is a patch to refer:

```[diff]
--- a/fs/namespace.c
+++ b/fs/namespace.c
@@ -1739,6 +1739,39 @@ static inline bool may_mandlock(void)
 }
 #endif
 
+static int can_umount(const struct path *path, int flags)
+{
+	struct mount *mnt = real_mount(path->mnt);
+
+	if (flags & ~(MNT_FORCE | MNT_DETACH | MNT_EXPIRE | UMOUNT_NOFOLLOW))
+		return -EINVAL;
+	if (!may_mount())
+		return -EPERM;
+	if (path->dentry != path->mnt->mnt_root)
+		return -EINVAL;
+	if (!check_mnt(mnt))
+		return -EINVAL;
+	if (mnt->mnt.mnt_flags & MNT_LOCKED) /* Check optimistically */
+		return -EINVAL;
+	if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))
+		return -EPERM;
+	return 0;
+}
+
+int path_umount(struct path *path, int flags)
+{
+	struct mount *mnt = real_mount(path->mnt);
+	int ret;
+
+	ret = can_umount(path, flags);
+	if (!ret)
+		ret = do_umount(mnt, flags);
+
+	/* we mustn't call path_put() as that would clear mnt_expiry_mark */
+	dput(path->dentry);
+	mntput_no_expire(mnt);
+	return ret;
+}
 /*
  * Now umount can handle mount points as well as block devices.
  * This is important for filesystems which use unnamed block devices.
```

**Review by Fede2782 — CHANGES_REQUESTED** [review]
Really good idea

**Review by Fede2782 — APPROVED** [review]

File: kernel/Makefile [hunk 2]
```diff
@@ -43,6 +43,10 @@ ifeq ($(shell grep -q "struct selinux_state " $(srctree)/security/selinux/includ
 ccflags-y += -DKSU_COMPAT_HAS_SELINUX_STATE
 endif
 
+ifeq ($(shell grep -q "int path_umount" $(srctree)/fs/namespace.c; echo $$?),0)
+ccflags-y += -DKSU_UMOUNT
+endif
```
Comment (tiann) [inline kernel/Makefile:48]: Add the warning here.

File: kernel/core_hook.c [hunk 3]
```diff
@@ -513,13 +513,15 @@ static bool should_umount(struct path *path)
 
 static void ksu_umount_mnt(struct path *path, int flags)
 {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 0)
```
Comment (tiann) [inline kernel/core_hook.c:516]: #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 0) || if defined(XXX)

This way it will never break gki.

File: kernel/core_hook.c [hunk 4]
```diff
@@ -513,13 +513,15 @@ static bool should_umount(struct path *path)
...
+#ifdef KSU_UMOUNT
 	int err = path_umount(path, flags);
 	if (err) {
 		pr_info("umount %s failed: %d\n", path->dentry->d_iname, err);
 	}
 #else
 	// TODO: umount for non GKI kernel
+	#warning Did you know you can backport path_umount to fs/namespace.c from 5.9?
```
Comment (tiann) [inline kernel/core_hook.c:523]: Remove this because the makefile will warn it.

File: kernel/setup.sh [hunk 5]
```diff
@@ -15,6 +15,8 @@ else
      exit 127
 fi
 
+grep -q "int path_umount" fs/namespace.c || echo "Did you know you can backport path_umount to fs/namespace.c from 5.9? Read: https://kernelsu.org/guide/how-to-integrate-for-non-gki.html#backport-path_umount-from-linux-59"
```
Comment (tiann) [inline kernel/setup.sh:18]: If you add the warning to Makefile, this can be removed.

**Review by tiann — CHANGES_REQUESTED** [review]

**Comment by tiann** [comment]
> ready to squash I guess. I'll squash after another approving review

You don't need to squash, i will make a squash merge.

**Comment by chiteroman** [comment]
+1 but why don't use latest code ?

https://elixir.bootlin.com/linux/latest/A/ident/path_umount

```cpp
/**
 * path_mounted - check whether path is mounted
 * @path: path to check
 *
 * Determine whether @path refers to the root of a mount.
 *
 * Return: true if @path is the root of a mount, false if not.
 */
static inline bool path_mounted(const struct path *path)
{
	return path->mnt->mnt_root == path->dentry;
}

static int can_umount(const struct path *path, int flags)
{
	struct mount *mnt = real_mount(path->mnt);

	if (!may_mount())
		return -EPERM;
	if (!path_mounted(path))
		return -EINVAL;
	if (!check_mnt(mnt))
		return -EINVAL;
	if (mnt->mnt.mnt_flags & MNT_LOCKED) /* Check optimistically */
		return -EINVAL;
	if (flags & MNT_FORCE && !capable(CAP_SYS_ADMIN))
		return -EPERM;
	return 0;
}

// caller is responsible for flags being sane
int path_umount(struct path *path, int flags)
{
	struct mount *mnt = real_mount(path->mnt);
	int ret;

	ret = can_umount(path, flags);
	if (!ret)
		ret = do_umount(mnt, flags);

	/* we mustn't call path_put() as that would clear mnt_expiry_mark */
	dput(path->dentry);
	mntput_no_expire(mnt);
	return ret;
}
```

**Comment by Fede2782** [comment]
> Is this attention grabbing enough? image
> 
> The thing is, the normal output is not even noticeable.

I would not waste time on this. Just write clearly in release notes and in the documentation (must be written clearly in both sections: krpobes and manual) and developers will notice. If some dev doesn't read the users will very likely complain to him and will implementation asap

**Comment by Fede2782** [comment]
For the docs i think is more coherent  this https://github.com/Fede2782/KernelSU/blob/main/website/docs/guide/how-to-integrate-for-non-gki.md

**Comment by ExtremeXT** [comment]
+1

**Comment by Zerototheleft122105** [comment]
Todo lo que proponéis me resulta complicado Pero llegará el momento de que lo domine. Un gran +1

File: kernel/Makefile [hunk 6]
```diff
@@ -22,6 +22,25 @@ else
...
+$(info -- ██▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▓██████████████████)
+$(info -- ██░░░░░░░░░░░░░░░▓██████████████████)
+$(info -- ██░░░░░░░░░░░░░░░▓██████████████████)
+$(info -- ██░░░░░░░░░░░░░░░▓██████████████████)
+$(info -- ██░░░░░░░░░░░░░░░▓██████████████████)
+$(info -- ██░░░░░░░░░░░░░░░▓██████████████████)
+$(info -- ██░░░░░░░░░░░░░░░▓██████████████████)
+$(info -- ██░░░░░░░░░░░░░░░▓██████████████████)
+$(info -- ████████████████████████████████████)
+
```
Comment (tiann) [inline kernel/Makefile:43]: please remove this noisy

File: kernel/Makefile [hunk 7]
```diff
@@ -62,6 +81,23 @@ $(info -- KernelSU Manager signature hash: $(KSU_EXPECTED_HASH))
...
+$(info -- / Did you know you can backport    \)
+$(info -- | path_umount to get module umount |)
+$(info -- \ working?                         /)
+$(info --  ----------------------------------)
+$(info --         \   ^__^)
+$(info --          \  (oo)\_______)
+$(info --             (__)\       $(CLOSE_PARENTHESES)\/\)
+$(info --                 ||----w |)
+$(info --                 ||     ||)
+$(info -- Read: https://kernelsu.org/guide/how-to-integrate-for-non-gki.html#path_umount)
```
Comment (tiann) [inline kernel/Makefile:98]: I don't want to open this file every time and see that most of the content is useless to me. Keep it simple please, one line is enough.

**Review by tiann — CHANGES_REQUESTED** [review]

**Review by tiann — APPROVED** [review]

**Review by nndmt034 — APPROVED** [review]
